cluster.go

```diff
@@ -0,0 +1,464 @@
+
+	// Create a map of known nodes
+	known := make(map[string]struct{}, len(c.addrs))
+	for _, addr := range c.addrs {
+		known[addr] = struct{}{}
+	}
+
+	// Populate slots, store unknown nodes
+	for _, info := range infos {
+		for i := info.Min; i <= info.Max; i++ {
+			c.slots[i] = info.Addrs
+		}
+
+		for _, addr := range info.Addrs {
+			if _, ok := known[addr]; !ok {
+				c.addrs = append(c.addrs, addr)
+				known[addr] = struct{}{}
+			}
+		}
+	}
+
+	// Shuffle addresses
+	for i := range c.addrs {
+		j := rand.Intn(i + 1)
+		c.addrs[i], c.addrs[j] = c.addrs[j], c.addrs[i]
+	}
+}
+
+//------------------------------------------------------------------------------
+
+var errNoAddrs = errors.New("redis: no addresses")
+
+type ClusterOptions struct {
+	// A seed-list of host:port addresses of known cluster nodes
+	Addrs []string
+
+	// An optional password
+	Password string
+
+	// The maximum number of MOVED/ASK redirects to follow, before
+	// giving up. Default: 16
+	MaxRedirects int
+
+	// The maximum number of open connections to cluster nodes. Default: 10.
+	// WARNING: Each connection maintains its own connection pool. The maximum
+	// possible number of socket connections is therefore MaxConns x PoolSize
+	MaxConns int
+
+	// The maximum number of TCP sockets per connection. Default: 5
+	PoolSize int
+
+	// Timeout settings
+	DialTimeout, ReadTimeout, WriteTimeout, IdleTimeout time.Duration
+}
+
+func (opt *ClusterOptions) getPoolSize() int {
+	if opt.PoolSize < 1 {
+		return 5
+	}
+	return opt.PoolSize
+}
+
+func (opt *ClusterOptions) getMaxConns() int {
+	if opt.MaxConns < 1 {
+		return 10
+	}
+	return opt.PoolSize
+}
+
+func (opt *ClusterOptions) getDialTimeout() time.Duration {
+	if opt.DialTimeout == 0 {
+		return 5 * time.Second
+	}
+	return opt.DialTimeout
+}
+
+func (opt *ClusterOptions) getMaxRedirects() int {
+	if opt.MaxRedirects < 1 {
+		return 16
+	}
+	return opt.MaxRedirects
+}
+
+func (opt *ClusterOptions) getAddrs() ([]string, error) {
+	if len(opt.Addrs) < 1 {
+		return nil, errNoAddrs
+	}
+	return opt.Addrs, nil
+}
+
+func (opt *ClusterOptions) ClientOptions(addr string) *Options {
+	return &Options{
+		Addr:     addr,
+		DB:       0,
+		Password: opt.Password,
+
+		DialTimeout:  opt.getDialTimeout(),
+		ReadTimeout:  opt.ReadTimeout,
+		WriteTimeout: opt.WriteTimeout,
+
+		PoolSize:    opt.getPoolSize(),
+		IdleTimeout: opt.IdleTimeout,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+const HashSlots = 16384
+
+// HashSlot returns a consistent slot number between 0 and 16383
+// for any given string key
+func HashSlot(key string) int {
+	if s := strings.IndexByte(key, '{'); s > -1 {
+		if e := strings.IndexByte(key[s+1:], '}'); e > 0 {
+			key = key[s+1 : s+e+1]
+		}
+	}
+	if key == "" {
+		return rand.Intn(HashSlots)
+	}
+	return int(crc16sum(key)) % HashSlots
+}
+
+// CRC16 implementation according to CCITT standards.
```

Comment:
Fixed in https://github.com/go-redis/redis/commit/4bbb0cf19944b2d56f6bb894eef0b1f91c7f5b53